File: src/classes/joueur.py
"""
Classe représentant un joueur
"""

from classes.carte import Carte

class Joueur :
    """
    Joueur du jeu Hit!
    """

    def __init__ (self, nom : str) -> None :
        self.nom :  str = nom
        self.score : int = 0
        self.carte_en_cours : list[Carte]= []
        self.cartes_sauvegardees : list[Carte] = []

    def __str__ (self) -> str :
        return f"Joueur {self.nom} (score : {self.score})"

    def __lt__ (self, autre) -> bool :
        if self.score < autre.score :
            return True
        return False

    def sauvegarder_cartes (self) -> None :
        """
        Permet à un joueur de sauvegarder les cartes qu'iel a conservé pendant un tour
        """
        score_actuel : int = self.score
        nombre_cartes : int = len(self.carte_en_cours)
        for carte in self.carte_en_cours :
            self.score += carte.get_valeur()
            self.cartes_sauvegardees.append(carte)
        self.carte_en_cours = []
        print(f"Le joueur {self.nom} a sauvegardé {nombre_cartes}\
              cartes pour une valeur de {self.score - score_actuel}.\n\
              Son score s'éléve maintenant à {self.score}")

File: C++/src/classes/Joueur.cpp
#include <time.h>
#include <string>
#include <iostream>

#include "Joueur.h"

Joueur::Joueur(const std::string nom) : nom(nom) {
    audace = (float)rand() / (float)RAND_MAX;
    attention = (float)rand() / (float)RAND_MAX;
}

Joueur::Joueur(const std::string nom, float audace, float attention) : nom(nom), audace(audace), attention(attention) {}

std::string Joueur::toString() const {
    return "Joueur " + nom + " (score : " + std::to_string(score) + ") [" + std::to_string(audace) + ", " + std::to_string(attention) + "]";
}

int Joueur::comparerScore(Joueur joueur2) {
    if (score < joueur2.getScore()) {
        return -1;
    }
    if (score == joueur2.getScore()) {
        return 0;
    }
    return 1;
}

int Joueur::getScore() const {return score;}

std::string Joueur::getNom() const {return nom;}

float Joueur::getAudace() const {return audace;}

float Joueur::getAttention() const {return attention;}

void Joueur::sauvegarderCartes() {
    int scoreActuel = score;
    int nombreCartes = cartesEnCours.size();
    for (Carte carte : cartesEnCours) {
        score += carte.getValeur();
        cartesSauvegardees.push_back(carte);
    }
    cartesEnCours.empty();
    if (nombreCartes == 0) {
        std::cout << "Le joueur " << nom << " n'a pas de carte à sauvegarder" << std::endl;
    }
    else {
        std::cout << "Le joueur " << nom << " a sauvegardé " << std::to_string(nombreCartes) <<
                     " carte(s) pour une valeur de " << std::to_string(score - scoreActuel) <<
                     " points.\nSon score s'eleve maintenant à " << std::to_string(score) << std::endl; 
    }
}

